contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -236,6 +242,126 @@ exec_tsql_stmt(PLtsql_execstate *estate, PLtsql_stmt *stmt, PLtsql_stmt *save_es
+		ereport(ERROR,
+			(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+				errmsg("User does not have permission to use the KILL statement")));
+	}
+
+	// Convert SPID string to int
+	// Although the T-SQL grammar only accepts a numeric value, and it must always be present, play it safe
+	Assert(stmt->spid != NULL);     
+	Assert(strlen(stmt->spid) > 0);
+
+	spid = (int) strtol(stmt->spid, &strtol_endptr, 10);
+	if ((spid == 0) && (*strtol_endptr != '\0'))
+	{
+		ereport(ERROR,
+			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				errmsg("Session ID %s is not valid", stmt->spid)));            
+	}
+	// SPID value must be positive; the T-SQL grammar allows only a non-negative nuber to be specified
+	// Yet, play it safe and test for it
+	// A variable or expression is not allowed and caught in the parser
+	// All other variants of T-SQL KILL are not supported, this is caught in the parser
+	if (spid <= 0)
+	{
+		ereport(ERROR,
+			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				errmsg("Session ID %d is not valid", spid)));
+	}
+
+	// Verify it is an actually existing process; otherwise we might just be killing any process on the host
+	proc = BackendPidGetProc(spid);
+	if (proc == NULL)
+	{
+		ereport(ERROR,
+			errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				errmsg("Process ID %d is not an active process ID", spid));
+	}
+
+	// Do not kill ourselves 
+	if (spid == MyProcPid)
+	{
+		ereport(ERROR,
+			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				errmsg("Cannot use KILL to kill your own process.")));
+	}		
+
+	// Verify this is a TDS connection, not a PG connection: we should not kill PG connections from T-SQL.
+	// This can be verified by checking the session to be present in sys.dm_exec_sessions or 
+	// sys.dm_exec_connections, which contains T-SQL connections only
+	// (unlike sys.syprocesses which also contains PG connections since this view is also 
+	//  based on pg_locks and pg_stat_activity)
+	{
+		char *tmpqry =
+			"SELECT DISTINCT 1 FROM sys.dm_exec_sessions WHERE session_id = %d " ;
+	
+		char *query = psprintf(tmpqry, spid);
+	
+		int rc = SPI_execute(query, true, 1);
+	
+		if (rc != SPI_OK_SELECT)
+			ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+					errmsg("SPI_execute failed: %s", SPI_result_code_string(rc))));
+
+		// 1 row found: TDS connection	 		
+		// 0 rows found: PG connection (since the connection was found to exist above)
+		if (SPI_processed == 0) 
+		{
+			ereport(ERROR,
+				(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+					errmsg("Process ID %d is not an active process ID for a TDS connection", spid)));
+		}
+
+		SPI_freetuptable(SPI_tuptable);		
+	}
+
+	// All validations passed, send the signal to the backend process
+	// This is basically the same as what pg_terminate_backend() does
+	if (kill(spid, SIGTERM))
+	{
+		ereport(WARNING,
+			(errmsg("Could not send signal to process %d: %m", spid)));
+	}
+
+	// Send no further message to the client, irrespective of the result
+
+	exec_set_rowcount(0);
```

Comment:
IIRC, This will only set the rowcount to 0 and we will be sending a TDS-Response with that values (0).
Is that the behaviour expected or we do not want to send the TDS-Response at all?

Reply:
The session executing KILL does not terminate. In SQL Server the rowcount is reset to 0, so this is the intended behaviour. Will add a comment to the code.